### PR TITLE
make `util::Progress` thread-safe as prerequisite of #11448

### PR DESCRIPTION
**Pull request by Byron**
This is a pre-requisite for obtaining progress information from `gitoxde` which is integrated [via this PR](https://github.com/rust-lang/cargo/pull/11448).

Obtaining progress information works by having a thread poll `gitoxide`'s progress in regular intervals.

### Alternatives Considered

`gitoxide` hands progress information to a `Progress` trait which allows it to build hierarchies of progress information to reflect potentially concurrent state updates. It also assumes that progress updates are fast and doesn't rate-limit its own calls.

Implementing this trait for `Progress` would have to involve rate limiting, which typically uses the local time to determine if a call will cause a display update or not, which at least in the past has been prohibitively slow. By now, at least on linux, I think this has been fixed but I am not sure about Windows.

As there are uncertainties to this approach and which also comes with considerable higher complexity (and more code), I decided to instead make `util::Progress` thread-safe with the minimal amount of changes.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -50,6 +50,7 @@ libgit2-sys = "=0.14.1"
 memchr = "2.1.3"
 opener = "0.5"
 os_info = "3.5.0"
+parking_lot = "0.12.1"
```
Comment (ehuss) [inline Cargo.toml:53]: Can you say why this needs parking_lot?
I'm generally reluctant to bring in new dependencies unless there is a strong reason. They increase build times, increase build and porting hazards, and potentially increase maintenance.

Comment (epage) [reply Cargo.toml:53]: From #11448

> It's both. I used std::sync::Mutex in an earlier iteration but found myself writing .unwrap() or .expect("thread didn't panic") which seemed worse than using parking_lot explicitly which is already part of the dependency tree.

Except I just looked and do not see `parking_lot` in the dependency tree

Comment (Byron) [reply Cargo.toml:53]: Sorry, I think this might have come up in the other PR but wasn't addressed.

I chose it to avoid dealing with lock poisoning. Not using it either meant that `fn -> T` has to become `fn -> Result<T>` where ever the lock is obtained, or that I'd have to use `unwrap()` or `expect("no panic in thread")`, both of which seemed like the something to avoid.

Is there other options you see, or a choice you would make from the alternatives presented here?

Comment (Byron) [reply Cargo.toml:53]: > Except I just looked and do not see parking_lot in the dependency tree

It's coming in with `gitoxide`.

File: src/cargo/util/config/mod.rs [hunk 2]
```diff
@@ -393,8 +394,13 @@ impl Config {
     }
 
     /// Gets a reference to the shell, e.g., for writing error messages.
-    pub fn shell(&self) -> RefMut<'_, Shell> {
-        self.shell.borrow_mut()
+    pub fn shell(&self) -> MutexGuard<'_, Shell> {
```
Comment (ehuss) [inline src/cargo/util/config/mod.rs:397]: Can the docstring here include some warnings about the restrictions the caller should be careful about? For example, this should not be called in the same thread if that thread already has a shell (otherwise it would deadlock).

Overall I'm a bit nervous about this since there won't be any compile time checks, and there are a *lot* of places that get a shell. I don't think I can review all of those call sites. I think the current code is probably ok, otherwise the RefCel would have panic'ed, but a panic is a lot better than a deadlock.

**Review by ehuss — COMMENTED** [review]
I'm wondering if an alternate approach would be to use a message queue? Cargo's own multi-threaded section does that (including handling progress). I'm not sure I follow the alternative discussion around a `Progress` trait. Does gitoxide generate progress updates at a very high rate?  Could Cargo's own `Throttle` type be used to help here?

I am a little concerned about the risk of deadlock, or future changes thinking it is OK to grab a `Shell` from a thread without considering what other threads are doing with the console output.  Perhaps that risk is low, but this is a little concerning.

Also, what have you done to test these changes? There are very few tests for the progress bar, so we need to do some manual verification.

File: Cargo.toml (hunk 1 above)
Comment (ehuss) [reply Cargo.toml:53]: Typically for dealing with that I just call `.unwrap()`. The chance of having a poisoned lock is small if you keep the locked region small, and avoid other panics within it. And the consequences are usually small, unless you try to lock a mutex in a `drop()` method, which I would usually not recommend (though it looks like this PR does that). 

It is certainly not ideal, and there has been a lot of discussion on whether or not to actually try to introduce a different API. There is risk with the design of parking_lot since invariants may no longer hold if some locked region only partially completed its work.

I think it should probably be OK to bring it in if it will be required by gitoxide.

File: src/cargo/util/progress.rs [hunk 3]
```diff
@@ -215,8 +218,8 @@ impl<'cfg> State<'cfg> {
         }
 
         // Only update if the line has changed.
-        if self.config.shell().is_cleared() || self.last_line.as_ref() != Some(&line) {
-            let mut shell = self.config.shell();
+        if self.shell.lock().is_cleared() || self.last_line.as_ref() != Some(&line) {
+            let mut shell = self.shell.lock();
```
[the remark this answers is not in the document]
Comment (Byron) [reply src/cargo/util/progress.rs]: That's a great catch, I have made the adjustment.

File: src/cargo/util/progress.rs [hunk 4]
```diff
@@ -229,15 +232,15 @@ impl<'cfg> State<'cfg> {
 
     fn clear(&mut self) {
         // No need to clear if the progress is not currently being displayed.
-        if self.last_line.is_some() && !self.config.shell().is_cleared() {
-            self.config.shell().err_erase_line();
+        if self.last_line.is_some() && !self.shell.lock().is_cleared() {
+            self.shell.lock().err_erase_line();
```
[the remark this answers is not in the document]
Comment (Byron) [reply src/cargo/util/progress.rs]: Thanks a lot, the adjustment was made.

File: src/cargo/util/config/mod.rs (hunk 2 above)
Comment (Byron) [reply src/cargo/util/config/mod.rs:401]: That's a good call, and I see the issue with going `RefCell` to `Mutex` as well. So much so that I'd also love to not have to do that.

For now I have added a big deadlock notice.

**Comment by Byron** [comment]
Thanks @ehuss for the review and suggestions - I will take another stab at a good alternative which is offloading the computation to a thread and keep the `ProgressBar` in the main thread, feeding it with messages. I am pretty sure I tried that but it failed because some other type couldn't be used in a thread that was needed.
However, I just looked at it again and believe that this dependency can be removed to make it work. I will give it another shot asap.

>  Does gitoxide generate progress updates at a very high rate? Could Cargo's own Throttle type be used to help here?

Yes, it will, but with `Throttle` it should be fine. Implementing custom `Progress` is doable, but the code will definitely be harder to maintain due to increased complexity.

To sum it up: I also think this change introduces some insecurity into the code-base that is best avoided, and it will be worth it to try harder to instead offload the `gitixode` operation into its own thread to have the main thread process progress instead. It should be possible, and I will be back with the results which hopefully leads to the closure of this PR.

**Comment by Byron** [comment]
It took me a while but I managed to manufacture a version that offloads the `gitoxide` fetch into a separate thread and processes the progress in the main thread 🎉. To me this seems like the solution I always wanted. The required modification will be put on top of the [integration PR](https://github.com/rust-lang/cargo/pull/11448).
